### PR TITLE
Error and warning improvements

### PR DESCRIPTION
**Pull request by jjmaldonis**
Summary:

This PR creates custom `Exception` objects when errors are thrown from the Deepgram SDK and provides support for the new `warnings` object in API responses.

---

Details:

Prior to this PR, Deepgram errors were distinguished by prefixing the error message with a `"DG:"` string. This PR removes the string prefix from all error messages and the exceptions were replaced with custom `Exception` objects. This allows users to use `try/except` blocks to handle errors from the Deepgram SDK, which is standard practice for Python libraries.

Two new exceptions will be thrown: `DeepgramSetupError` will be thrown if an error occurs while instantiating the Deepgram SDK (such as a missing API key), and `DeepgramApiError` will be thrown when errors are encountered during API requests. Both errors inherit from a `DeepgramError` base class exception, which allows users to catch all Deepgram exceptions in one `except` block.

The `DeepgramApiError` has a number of fields the user may find useful when handling this exception, such as an HTTP status code if one exists.

The new support for warnings defaults to printing a warning to `stdout` using Python's built-in `warnings` module. This is standard practice for handling warnings.

Two options were added to allow users to customize how warnings are handled. The `Options` class that is passed to the Deepgram SDK duration initialization has two new fields: `suppress_warnings` which prevents the warnings from being printed to the console, and `raise_warnings_as_errors` which raises warnings as errors if a warning is found in the API response.

---

Testing:

- The `pytest` tests run successfully on Python 3.11 and Python 3.9. I did not test on Python 3.10 but there are no changes that should impact it.

- The new support for warnings can be tested by hitting the Deepgram API with `summarization=true&detect_language=true` and using a non-English language. [Here is the German audio file I used for testing](https://github.com/deepgram/deepgram-python-sdk/files/12198647/german.zip).
  - I tested `suppress_warnings` and `raise_warnings_with_errors` with this API call and they work correctly.

- To test the new error handling, I used `tier=nova&language=de` (with any audio file/URL). It works well. There are errors I wasn't able to reproduce, such as 503 errors from Deepgram.

- I also tested timeout errors, which raise the appropriate exception (no change).

### CODE REVIEW

File: deepgram/errors.py
Comment (jkroll-deepgram) [inline deepgram/errors.py:38]: Does this mean we are presenting all of these errors as a type of Deepgram error? Would we maybe want other non-DG errors to bubble up rather than wrapping them as our own?

File: deepgram/errors.py
Comment (jkroll-deepgram) [inline deepgram/errors.py:64]: You're referring to `args[0]` a lot here, it would be helpful for readability to give that a named variable so it's clearer what that arg is.

File: deepgram/__init__.py
Comment (jjmaldonis) [reply deepgram/__init__.py:27]: I completely agree. The existing code has a mix of single quotes and double quotes. Python's standard is double quotes, so that's what I use. I did not want to add a ton of formatting changes to this PR, so I kept them minimal. I can change this one in particular because you raised it, but I'd like to go through and reformat the entire codebase in a later PR.

File: deepgram/__init__.py
Comment (jjmaldonis) [reply deepgram/__init__.py:18]: The line you highlighted is not an Auth error and instead implies that the user passed the incorrect type to the constructor. We could also throw a `TypeError` here, but `DeepgramSetupError` keeps the error codes consistent. Also `DeepgramSetupError` inherits from `ValueError` which is applicable when the user passes the wrong value/type to the constructor.

File: deepgram/_types.py
Comment (jjmaldonis) [reply deepgram/_types.py:31]: `TypedDict`s do not have default values, and none of the other classes in this file have default values. Technically these types should have `Optional` around them, but very few of the other optional fields in this file use `Optional` so I left it off here as well. This is something we can clean up in the future.

We could also convert the TypedDicts to dataclasses (I prefer dataclasses to TypedDicts). Alternatively, we could use TypedDict's `Required` and `NotRequired` types defined in this pep: https://peps.python.org/pep-0655/

File: deepgram/errors.py
Comment (jjmaldonis) [reply deepgram/errors.py:38]: In the previous iteration of the code, the Python SDK was catching all non-DG errors and returning them as `Exception`, sometimes without using the `from` operator (!!!). That means the non-DG errors - and the information available in them - were inaccessible.

I agree that we could allow these errors to bubble up on their own rather than catch them and wrap them in a Deepgram error. It would allow users to be more explicit about how to handle errors, but it would also make error handling more difficult for the user because they would need to handle multiple exception types.

The two approaches are targeted at different audiences: The implementation where errors are wrapped up into a single DG exception makes it easy to do simply error handling and therefore makes it easy for newer programmers to handle failed API requests. On the other hand, the implementation where errors bubble up requires the user to be explicit about how they handle the different use cases, which is likely more useful for power users.

The code I wrote grabs the underlying non-DG error and adds it as a field to the DG error. I hope this satisfies both user groups because: the programmers with less experience can simply catch one error, and the power users can grab the underlying HTTP error and perform more advanced error handling.

Aside: I said, "the implementation where errors bubble up requires the user to be explicit about how they handle the different use cases". That isn't completely true because users could always write `except Exception:` to catch _all_ errors, but I try to highly discourage the use of `except Exception`. It is almost always too broad. For example, if a user passes an incorrect type that results in a `TypeError`, they wouldn't want their retry logic to trigger and keep making API calls with the same `TypeError`.

After that long explanation, I hope this code serves both new users and power users by making it easy to handle failed API requests in both simple and complex ways.

File: deepgram/errors.py
Comment (jjmaldonis) [reply deepgram/errors.py:64]: Thanks, fixed. I created a new variable: `error_or_warning_data = args[0]`

File: deepgram/_types.py
Comment (jkroll-deepgram) [reply deepgram/_types.py:31]: I don't have a strong preference - do you think TypedDicts are marginally more performant, but dataclasses are more flexible? We do have an awful lot of TypedDicts already, so it makes sense to defer to that standard for now.

My interest in an explicit default value is mainly to make clear what the behavior will be when the user doesn't set anything. But I guess in this case, it's logical for warnings to be warnings by default... rather than being suppressed or errors.

File: deepgram/_utils.py
Comment (jkroll-deepgram) [reply deepgram/_utils.py:185]: The downside I see with the `in` syntax is that it gets verbose and repetitive for nested dictionaries. But you're only going 2 levels deep so it's not too bad. I'm fine with it as-is and I don't know of a specific style guide to be followed here. (Seems like the absence of a DG Python style guide is becoming a theme in this PR discussion!)

**Review by jkroll-deepgram — DISMISSED** [review]

**Comment by jkroll-deepgram** [comment]
Approved! But best to wait until signoff from Luke/DX before merging 🙏

**Comment by jjmaldonis** [comment]
> Oh one more question - can you add some tests in `tests/`, such as to assert that certain exceptions are thrown when expected? Or that warnings are suppressed/raised to errors when expected?

I added two tests that ensure the appropriate errors are raised.

**Review by jkroll-deepgram — APPROVED** [review]

**Pull request merged**
